qiskit/circuit/library/n_local/qaoa_ansatz.py

```diff
@@ -0,0 +1,257 @@
+                                     "the number of qubits of the cost operator {}"
+                                     .format(self._mixer.num_qubits, self._num_qubits))
+
+        return valid
+
+    def _build(self) -> None:
+        """Build the circuit."""
+        if self._data:
+            return
+
+        self._check_configuration()
+        self._data = []
+
+        # calculate bounds, num_parameters, mixer
+        self._calculate_parameters()
+
+        # parametrize circuit and build it
+        param_vector = ParameterVector("θ", self._num_parameters)
+        circuit = self._construct_circuit(param_vector)
+
+        # append(replace) the circuit to this
+        self.add_register(circuit.num_qubits)
+        self.compose(circuit, inplace=True)
+
+    @property
+    def parameters(self) -> Set[Parameter]:
+        """Get the :class:`~qiskit.circuit.Parameter` objects in the circuit.
+
+        Returns:
+            A set containing the unbound circuit parameters.
+        """
+        self._build()
+        return super().parameters
+
+    def _calculate_parameters(self):
+        """Calculated internal parameters of the circuit to be built."""
+        from qiskit.opflow import OperatorBase
+        if isinstance(self._mixer, QuantumCircuit):
+            self._num_parameters = (1 + self._mixer.num_parameters) * self._reps
+            self._bounds = [(None, None)] * self._reps + \
+                           [(None, None)] * self._reps * self._mixer.num_parameters
+        elif isinstance(self._mixer, OperatorBase):
+            self._num_parameters = 2 * self._reps
+            self._bounds = [(None, None)] * self._reps + [(None, None)] * self._reps
+        elif self._mixer is None:
+            self._num_parameters = 2 * self._reps
+            self._bounds = [(None, None)] * self._reps + [(0, 2 * np.pi)] * self._reps
+
+    def _construct_circuit(self, parameters) -> QuantumCircuit:
+        """Construct a parametrized circuit."""
+        if not len(parameters) == self._num_parameters:
+            raise ValueError('Incorrect number of angles: expecting {}, but {} given.'.format(
+                self._num_parameters, len(parameters)
+            ))
+
+        # local imports to avoid circular imports
+        from qiskit.opflow import CircuitStateFn
+        from qiskit.opflow import CircuitOp, EvolutionFactory
+
+        circuit_op = CircuitStateFn(self.initial_state)
+
+        # iterate over layers
+        for idx in range(self._reps):
+            # the first [:self._reps] parameters are used for the cost operator,
+            # so we apply them here
+            circuit_op = (self._cost_operator * parameters[idx]).exp_i().compose(circuit_op)
+            from qiskit.opflow import OperatorBase
+            mixer = self.mixer_operator
+            if isinstance(mixer, OperatorBase):
+                mixer = cast(OperatorBase, mixer)
```

Comment:
Since we don't have mypy checks we could also remove these casts

Reply:
Yes, but without these casts PyCharm highlights "unresolved ....", so I'd prefer to keep them. Type hints in the IDE again...